tiledb/stdx/synchronized_optional/test/unit_synchronized_optional.cc

```diff
@@ -0,0 +1,360 @@
+/**
+ * @file tiledb/stdx/synchronized_optional/test/unit_synchronized_optional.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2022 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#define CATCH_CONFIG_MAIN
+#include "../synchronized_optional.h"
+#include "test/support/tdb_catch.h"
+
+using stdx::synchronized_optional;
+using soint = synchronized_optional<int>;
+
```

Comment:
My main comment about the tests is that there should be some tests that exercise asynchronous access to `std::synchronized_optional`.  It seems to be possible to get races and deadlock (though perhaps something is wrong with my tests).

Reply:
Testing this well needs C++20 `std::barrier` or equivalent to get deterministic tests. It would be possible to do some stress-testing on it, but at this point I'd prefer to wait and do a proper job of it.